text/0626-invoke-helper.md

```diff
@@ -142,124 +118,252 @@ function myProgram(data = []) {
+the data fetching functionality, one to handle updating the fetch, and one to
+handle tearing it down, because those different pieces of code need to run at
+different portions of the component lifecycle. If we want to reuse these
+functions elsewhere, this adds a lot of boilerplate to integrate the functions
+in each lifecycle hook.
+
+There are a few alternatives that would allow us to extract this functionality
+together.
+
+1. We could use mixins, since they allow us to specify multiple functions and
+   mix them into a class. Mixins however introduce a lot of complexity in the
+   inheritance hierarchy and are considered an antipattern, so this is not a
+   good choice overall.
+
+2. We could extract the functionality out to separate components. Components
+   have a contained lifecycle, so they can manage any piece of functionality
+   completely in isolation. This works nicely for the document title, but adds
+   a lot of complexity for the data fetching, since we need to yield the data
+   back out via the template:
+
+    ```js
+    // app/components/doc-title.js
+    import Component from '@ember/component';
+
+    export default class DocTitle extends Component {
+      didReceiveAttrs() {
+        document.title = this.title;
+      }
+    }
+    ```
+    ```js
+    // app/components/fetch-data.js
+    import Component from '@ember/component';
+    import { fetch, cancelFetch } from 'fetch-with-cancel';
+
+    export default class FetchData extends Component {
+      @tracked data;
+
+      didReceiveAttrs() {
+        // Cancel the previous fetch if it's still running
+        cancelFetch(this.promise);
+
+        // create a new fetch request, and set the data property to the response
+        this.promise = fetch(this.url)
+          .then((response) => response.json());
+          .then((data) => this.data = data);
+      }
+
+      willDestroy() {
+        cancelFetch(this.promise);
+      }
+    }
+    ```
+    ```hbs
+    <!-- app/components/fetch-data.hbs -->
+    {{yield this.data}}
+    ```
+    ```hbs
+    <!-- app/components/search.hbs -->
+    <DocTitle @title='Search Result for "{{@query}}"'/>
+
+    <FetchData @url="www.example.com/search?query={{@query}}" as |data|>
+      ...
+    </FetchData>
+    ```
+
+   This structure is also not ideal because the components aren't being used for
+   templating, they're just being used for logic effectively. So, there's no
+   reason to add the overhead of a component here.
+
+3. We could use other template constructs, such as helpers or modifiers. Both
+   helpers and modifiers have lifecycles, like components, and can be used to
+   contain functionality. Modifiers aren't really a good choice here though,
+   because it would require us to add an element that we don't need. So, helpers
+   are the better option.
+
+   Helpers work, but like components they require us to move some of our logic
+   into the template, even if that isn't really necessary:
```

Comment:
I'm not sure I follow this part. What is the unnecessary part in the template?

Reply:
Right, exactly. You can't refer to the value of the helper in JS anymore, so any computed properties you have that may rely on the value of the helper now need to be in the template, any other business logic now needs to be moved into the template, etc.

Reply:
Hmm, I don't see that as unnecessary 🤔 

Either way (either with this RFC or with templates) you have to invoke the helper. The only question is where that invocation actually is.